frontend/src/app/main/data/workspace/thumbnails.cljs

```diff
@@ -0,0 +1,174 @@
+  [[event [old-objects new-objects]]]
+  (let [changes (-> event deref :changes)
+
+        extract-ids
+        (fn [{type :type :as change}]
+          (case type
+            :add-obj [(:id change)]
+            :mod-obj [(:id change)]
+            :del-obj [(:id change)]
+            :reg-objects (:shapes change)
+            :mov-objects (:shapes change)
+            []))
+
+        get-frame-id
+        (fn [id]
+          (let [shape (or (get new-objects id)
+                          (get old-objects id))]
+
+            (or (and (= :frame (:type shape)) id)
+                (:frame-id shape))))
+
+        ;; Extracts the frames and then removes nils and the root frame
+        xform (comp (mapcat extract-ids)
+                    (map get-frame-id)
+                    (remove nil?)
+                    (filter #(not= uuid/zero %))
+                    (filter #(contains? new-objects %)))]
+
+    (into #{} xform changes)))
+
+(defn thumbnail-change?
+  "Checks if a event is only updating thumbnails to ignore in the thumbnail generation process"
+  [event]
+  (let [changes (-> event deref :changes)
+
+        is-thumbnail-op?
+        (fn [{type :type attr :attr}]
+          (and (= type :set)
+               (= attr :thumbnail)))
+
+        is-thumbnail-change?
+        (fn [change]
+          (and (= (:type change) :mod-obj)
+               (->> change :operations (every? is-thumbnail-op?))))]
+
+    (->> changes (every? is-thumbnail-change?))))
+
+(defn watch-state-changes
+  "Watch the state for changes inside frames. If a change is detected will force a rendering
+  of the frame data so the thumbnail can be updated."
+  []
+  (ptk/reify ::watch-state-changes
+    ptk/WatchEvent
+    (watch [_ _ stream]
+      (let [stopper (->> stream
+                         (rx/filter #(or (= :app.main.data.workspace/finalize-page (ptk/type %))
+                                         (= ::watch-state-changes (ptk/type %)))))
+
+            objects-stream (->> (rx/concat
+                                 (rx/of nil)
+                                 (rx/from-atom refs/workspace-page-objects {:emit-current-value? true}))
+                                ;; We need to keep the old-objects so we can check the frame for the
+                                ;; deleted objects
+                                (rx/buffer 2 1))
+
+            frame-changes (->> stream
+                               (rx/filter dch/commit-changes?)
+
+                               ;; Async so we wait for additional side-effects of commit-changes
+                               (rx/observe-on :async)
+                               (rx/filter (comp not thumbnail-change?))
+                               (rx/with-latest-from objects-stream)
+                               (rx/map extract-frame-changes)
+                               (rx/share))]
+
+        (->> frame-changes
+             (rx/take-until stopper)
```

Comment:
take-until, always on the end of the stream

Reply:
ok